virttest/utils_net.py

```diff
@@ -437,6 +437,56 @@ def is_up(self):
         else:
             return False
 
+    @warp_init_del
+    def promisc_on(self):
+        '''
+        Enable promiscuous mode on the interface.
+        Equivalent to ip link set [iface] promisc on.
+        '''
+
+        # Get existing device flags
+        ifreq = struct.pack('16sh', self.name.encode(), 0)
+        flags = struct.unpack('16sh',
+                              fcntl.ioctl(sockfd, arch.SIOCGIFFLAGS, ifreq))[1]
+
+        # Set new flags
+        flags = flags | arch.IFF_PROMISC
+        ifreq = struct.pack('16sh', self.name.encode(), flags)
+        fcntl.ioctl(sockfd, arch.SIOCSIFFLAGS, ifreq)
+
+    @warp_init_del
+    def promisc_off(self):
+        '''
+        Disable promiscuous mode on the interface.
+        Equivalent to ip link set [iface] promisc off.
+        '''
+
+        # Get existing device flags
+        ifreq = struct.pack('16sh', self.name.encode(), 0)
+        flags = struct.unpack('16sh',
+                              fcntl.ioctl(sockfd, arch.SIOCGIFFLAGS, ifreq))[1]
+
+        # Set new flags
+        flags = flags & ~arch.IFF_PROMISC
+        ifreq = struct.pack('16sh', self.name.encode(), flags)
+        fcntl.ioctl(sockfd, arch.SIOCSIFFLAGS, ifreq)
+
+    @warp_init_del
+    def is_promisc(self):
+        '''
+        Return True if the interface promiscuous mode is on, False otherwise.
+        '''
+        # Get existing device flags
+        ifreq = struct.pack('16sh', self.name.encode(), 0)
+        flags = struct.unpack('16sh',
+                              fcntl.ioctl(sockfd, arch.SIOCGIFFLAGS, ifreq))[1]
+
+        # Set new flags
+        if flags & arch.IFF_PROMISC:
+            return True
+        else:
+            return False
+
```

Comment:
This seems pretty much alike with the function `up` `down` `is_up`. Can we reorganize them like:
```
def set_device_flag(self, flag='IFF_UP', on=True):
    ifreq = struct.pack('16sh', self.name.encode(), 0)
    flags = struct.unpack('16sh',
                          fcntl.ioctl(sockfd, arch.SIOCGIFFLAGS, ifreq))[1]
    if on:
        flags = flags | getattr(arch, flag)
    else:
        flags = flags & ~getattr(arch, flag)
    ifreq = struct.pack('16sh', self.name.encode(), flags)
    fcntl.ioctl(sockfd, arch.SIOCSIFFLAGS, ifreq)

def is_device_flag_on(self, flag=IFF_UP):
    ifreq = struct.pack('16sh', self.name.encode(), 0)
    flags = struct.unpack('16sh',
                          fcntl.ioctl(sockfd, arch.SIOCGIFFLAGS, ifreq))[1]

    if flags & getattr(arch, flag):
        return True
    else:
        return False

def up(self):
    self.set_device_flag()

def down(self):
    self.set_device_flag(on=False)

def promisc_on(self)
    self.set_device_flag(IFF_PROMISC)

def promisc_off(self)
    self.set_device_flag(IFF_PROMISC, on=False)
```

Reply:
Agree with the opinion above.